### PR TITLE
implement protoc-gen-orm

### PR DESCRIPTION
**Pull request by mirshahriar**

### CODE REVIEW

File: cmd/protoc-gen-orm/Makefile [hunk 1]
```diff
@@ -0,0 +1,8 @@
+gen:
+	make -C ../../protobuf/
+	@go install ./...
+
+move: gen
+	sudo mv ${GOPATH}/bin/protoc-gen-orm /usr/local/bin/
```
Comment (sadlil) [inline cmd/protoc-gen-orm/Makefile]: You don't need to move the protoc-gen-orm as go install already creates it in go/bin, that is in PATH.

File: cmd/protoc-gen-orm/Makefile [hunk 2]
```diff
@@ -0,0 +1,8 @@
+gen:
+	make -C ../../protobuf/
+	@go install ./...
```
Comment (sadlil) [inline cmd/protoc-gen-orm/Makefile]: Use this `go install` in `make install` in outer make. Do not need to create this make file.

File: cmd/protoc-gen-orm/main.go [hunk 3]
```diff
@@ -1,7 +1,49 @@
...
+	"os"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/gophersbd/ormpb/protoc-gen-orm/generator"
+)
 
 func main() {
-	fmt.Println("Hello I am going to be ORM. Please Code me up")
+
+	g := generator.New()
```
Comment (sadlil) [inline cmd/protoc-gen-orm/main.go]: Please make this main as small as possible. Move most of the code into functions and write unit test. Some build will not pass if the test coverage is less than 80%

File: protoc-gen-orm/generator/generator.go [hunk 4]
```diff
@@ -0,0 +1,2884 @@
...
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```
Comment (sadlil) [inline protoc-gen-orm/generator/generator.go]: Careful about the licence?

File: protoc-gen-orm/generator/generator.go [hunk 5]
```diff
@@ -0,0 +1,2884 @@
...
+	Init(g *Generator)
+	// Generate produces the code generated by the plugin for this file,
+	// except for the imports, by calling the generator's methods P, In, and Out.
+	Generate(file *FileDescriptor)
+	// GenerateImports produces the import declarations for this file.
+	// It is called after Generate.
+	GenerateImports(file *FileDescriptor)
+}
+
+var plugins []Plugin
```
Comment (sadlil) [inline protoc-gen-orm/generator/generator.go]: I can look into more about the implementation. But i relive maintaining your own plugin registry will not help. We are a plugin in this case. Our code will run with something like `--ormpb_out=` just like as `grpc-gateway`. We will implement the interface for the go plugin. Not introduce another Plugin type.

**Review by sadlil — CHANGES_REQUESTED** [review]
I will look more into this, about the generator implementation.

File: protoc-gen-orm/testdata/Makefile [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# Variables
+PROJECT = ormpb
+REPO_ROOT = ${GOPATH}/src/github.com/gophersbd
+ROOT = ${REPO_ROOT}/${PROJECT}
+
+regenerate:
+	@echo WARNING! THIS RULE IS PROBABLY NOT RIGHT FOR YOUR INSTALLATION
+	@prototool compile message_options.proto
+	@protoc --go_out=../../. -I${ROOT} ${ROOT}/protoc-gen-orm/testdata/message_options.proto
```
Comment (mirshahriar) [inline protoc-gen-orm/testdata/Makefile]: This should be like bellow:
`@protoc --go_out=plugins=orm:../../. -I${ROOT} ${ROOT}/protoc-gen-orm/testdata/message_options.proto`

Failed to make it works. Added another command to generate orm.

File: cmd/protoc-gen-orm/main.go [hunk 7]
```diff
@@ -1,7 +1,73 @@
...
-	fmt.Println("Hello I am going to be ORM. Please Code me up")
+	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
+	pflag.Parse()
+	// Convinces goflags that we have called Parse() to avoid noisy logs.
+	flag.CommandLine.Parse([]string{})
+
+	Start()
+}
+
+func Start() {
```
Comment (sadlil) [inline cmd/protoc-gen-orm/main.go]: Write test for Start Methods, may be break it down to multiple methods and add tests for each. I do really care about having tests

File: pkg/generator/template.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+var (
+	ormTemplate = template.Must(template.New("orm").Parse(`
+// Code generated by protoc-gen-orm. DO NOT EDIT.
+// source: {{ .GetName }}
+
+package {{ .GoPkg.Name }}
+
+const(
```
Comment (sadlil) [inline pkg/generator/template.go]: Those constants values are jeneric? isn't it? May be move those to a `pkg/runtime/tags` package and import things from there??

**Review by sadlil — CHANGES_REQUESTED** [review]

**Comment by sadlil** [comment]
@aerokite i have pushed some changes and also added some comments, can address about that?

File: cmd/protoc-gen-orm/main.go (hunk 7 above)
Comment (mirshahriar) [reply cmd/protoc-gen-orm/main.go]: Everything inside this Start() method has already been tested. Covered in `registry_test` and `generator_test`. If still necessary, I will just copy one TestMethod here

File: pkg/generator/template.go (hunk 8 above)
Comment (mirshahriar) [reply pkg/generator/template.go]: Yes. Right. OtherWise, This may conflict

**Review by sadlil — APPROVED** [review]

**Pull request merged**